### PR TITLE
Add new version private transaction on Elysium

### PR DESCRIPTION
**Pull request by thebevrishot**
## PR intention
Add Lelantus on the token layer so users can mint and spend. And token owners can control Lelantus status to enable or disable the feature.

## Code changes brief
- Updated RPCs to create tokens to allow token creators to specify Lelantus status
- Added RPCs and logic to mint and spend token using Lelantus protocol and added tests for them.
- Created Lelantus state on Elysium to manage anon groups and serial and added tests for them.
- Added Lelantus wallet on Elysium to manage coins and credentials.
- Added an RPC and logic to manage Lelantus status by token owner.
- Bumped `MAX_OP_RETURN_RELAY` to 16k to allow spend tx to spend more than one coin.
- Added an RPC to recover coins (`elysium_recoverlelantusmint`) using passphrases

### CODE REVIEW

**Comment by det789** [comment]
2  src/Makefile.elysium.include 
@@ -30,6 +30,7 @@ ELYSIUM_H = \
Multi-line comments are here!

You can now comment on multiple lines. Just click and drag on the  button.

Demonstrating selecting multiple lines for commenting

  elysium/sigma.h \
  elysium/sigmaprimitives.h \
  elysium/sigmadb.h \
  elysium/lelantusdb.h \
  elysium/signaturebuilder.h \
  elysium/sp.h \
  elysium/sto.h \
@@ -74,6 +75,7 @@ ELYSIUM_CPP = \
  elysium/sigma.cpp \
  elysium/sigmaprimitives.cpp \
  elysium/sigmadb.cpp \
  elysium/lelantusdb.cpp \
  elysium/signaturebuilder.cpp \
  elysium/sp.cpp \
  elysium/sto.cpp \
 1  src/Makefile.elysiumtest.include 
@@ -34,6 +34,7 @@ ELYSIUM_TEST_CPP = \
  elysium/test/sp_tests.cpp \
  elysium/test/strtoint64_tests.cpp \
  elysium/test/swapbyteorder_tests.cpp \
  elysium/test/lelantusdb_tests.cpp \
  elysium/test/tally_tests.cpp \
  elysium/test/uint256_extensions_tests.cpp \
  elysium/test/utils_tx.cpp
 1  src/dbwrapper.cpp 
@@ -126,6 +126,7 @@ CDBIterator::~CDBIterator() { delete piter; }
bool CDBIterator::Valid() { return piter->Valid(); }
void CDBIterator::SeekToFirst() { piter->SeekToFirst(); }
void CDBIterator::Next() { piter->Next(); }
void CDBIterator::Prev() { piter->Prev(); }

namespace dbwrapper_private {

 2  src/dbwrapper.h 
@@ -162,6 +162,8 @@ class CDBIterator

    void Next();

    void Prev();

    template<typename K> bool GetKey(K& key) {
        leveldb::Slice slKey = piter->key();
        try {
 118  src/elysium/lelantusdb.cpp 
@@ -7,68 +7,116 @@

namespace elysium {

enum class KeyType : uint8_t
{
    // mint
    Coin = 0, // 0<id><order> = <public coin>
    CoinIndex = 1, // 1<id><block> = <order_i><order_j>
    BlockStoringIndex = 2, // 2<order> = 1<id><block>
    GroupIndex = 3, // 3<id><group_id> = <block_i><block_j>

    // serials
    Serial = 4, // 4<id><serial> = <block>
    SerialStoringOrder = 5, // 5<order> = <id><serial>
    SerialLastOrder = 6,
};

void SafeSeekToPreviousKey(leveldb::Iterator *it, const leveldb::Slice& key)
{
    it->Seek(key);
    if (it->Valid()) {
        it->Prev();
    } else {
        it->SeekToLast();
    }
}
static const char DB_SERIAL             = 0x00;
static const char DB_SERIAL_SEQUENCE    = 0x01;
static const char DB_SERIAL_NEXT        = 0x02;

// DB
LelantusDB::LelantusDB(size_t nCacheSize, bool fMemory, bool fWipe) :
LelantusDb::LelantusDb(size_t nCacheSize, bool fMemory, bool fWipe) :
    db(fMemory ? "" : GetDataDir() / "elysium_lelantusdb", nCacheSize, fMemory, fWipe)
{
}

bool LelantusDB::HasSerial(Scalar const &serial)
bool LelantusDb::HasSerial(PropertyId id, Scalar const &serial)
{
    return db.Exists(std::make_tuple(DB_SERIAL, id, serial));
}

bool LelantusDB::RemoveSerials(int block)
bool LelantusDb::RemoveSerials(int block)
{
    auto sequence = ReadNextSerialSequence();
    if (sequence <= 0) {
        return false;
    }

    auto it = std::unique_ptr<CDBIterator>(db.NewIterator());
    auto _sequence = sequence - 1;
    swapByteOrder(_sequence);

    it->Seek(std::make_tuple(DB_SERIAL_SEQUENCE, _sequence));

    std::tuple<PropertyId, secp_primitives::Scalar, int> val;

    std::vector<std::tuple<char, uint64_t>> toBeRemoveSerialOrders;
    std::vector<std::tuple<char, PropertyId, secp_primitives::Scalar>> toBeRemoveSerials;

    CDataStream keyStream(SER_DISK, CLIENT_VERSION);
    for (; it->Valid()
        && (keyStream = it->GetKey()).size() > 0
        && keyStream[0] == DB_SERIAL_SEQUENCE;
        it->Prev()) {

        if (!it->GetValue(val)) {
            throw std::runtime_error("Fail to get value");
        }

        if (std::get<2>(val) < block) {
            break;
        }

        toBeRemoveSerialOrders.emplace_back();
        if (!it->GetKey(toBeRemoveSerialOrders.back())) {
            throw std::runtime_error("Fail to get sequence key");
        }

        toBeRemoveSerials.emplace_back(DB_SERIAL, std::get<0>(val), std::get<1>(val));

        sequence--;
    }

    for (auto const &removeKey : toBeRemoveSerialOrders) {
        if (!db.Erase(removeKey)) {
            throw std::runtime_error("Fail to erase key");
        }
    }

    for (auto const &removeKey : toBeRemoveSerials) {
        if (!db.Erase(removeKey)) {
            throw std::runtime_error("Fail to erase key");
        }
    }

    WriteNextSerialSequence(sequence);

    return true;
}

bool LelantusDB::WriteSerials(
bool LelantusDb::WriteSerials(
    int block,
    std::vector<std::pair<PropertyId, std::vector<Scalar>>> const &serials)
{
    auto order = ReadSerialOrder();
    auto sequence = ReadNextSerialSequence();

    for (auto const &propertySerials : serials) {
        // db.Write(std::make_tuple(), );
        for (auto const &serial : propertySerials.second) {
            auto _sequence = sequence++;
            swapByteOrder(_sequence);

            db.Write(std::make_tuple(DB_SERIAL_SEQUENCE, _sequence),
                std::make_tuple(propertySerials.first, serial, block));

            db.Write(std::make_tuple(DB_SERIAL, propertySerials.first, serial), 0);
        }
    }

    WriteNextSerialSequence(sequence);

    return true;
}

bool LelantusDB::WriteSerialOrder(uint64_t order)
bool LelantusDb::WriteNextSerialSequence(uint64_t sequence)
{
    return db.Write(KeyType::SerialLastOrder, order);
    return db.Write(DB_SERIAL_NEXT, sequence);
}

uint64_t LelantusDB::ReadSerialOrder()
uint64_t LelantusDb::ReadNextSerialSequence()
{
    uint64_t order;
    if (!db.Read(KeyType::SerialLastOrder, order)) {
    uint64_t sequence;
    if (!db.Read(DB_SERIAL_NEXT, sequence)) {
        return 0;
    }

    return order;
    return sequence;
}

} // namespace elysium 
 12  src/elysium/lelantusdb.h 
@@ -13,25 +13,25 @@

namespace elysium {

class LelantusDB
class LelantusDb
{
private:
    CCriticalSection cs;
    CDBWrapper db;

public:
    LelantusDB(size_t nCacheSize, bool fMemory = false, bool fWipe = false);
    LelantusDb(size_t nCacheSize, bool fMemory = false, bool fWipe = false);

public:
    bool HasSerial(Scalar const &serial);
    bool HasSerial(PropertyId id, Scalar const &serial);
    bool RemoveSerials(int block);
    bool WriteSerials(
        int block,
        std::vector<std::pair<PropertyId, std::vector<Scalar>>> const &serials);

private:
    bool WriteSerialOrder(uint64_t);
    uint64_t ReadSerialOrder();
protected:
    bool WriteNextSerialSequence(uint64_t);
    uint64_t ReadNextSerialSequence();
};

} // namespace elysium
 105  src/elysium/test/lelantusdb_tests.cpp 
@@ -0,0 +1,105 @@
#include "../lelantusdb.h"

#include "../../test/fixtures.h"
#include "../../test/test_bitcoin.h"

#include <boost/test/unit_test.hpp>

namespace elysium {

namespace {

class TestLelantusDb : public LelantusDb
{
public:
    TestLelantusDb() : LelantusDb(100)
    {
    }

// proxy
public:
    uint64_t ReadNextSerialSequence()
    {
        return LelantusDb::ReadNextSerialSequence();
    }
};

class LelantusDbTestingSetup : public TestingSetup
{
public:
    LelantusDbTestingSetup() : TestingSetup(CBaseChainParams::REGTEST)
    {
    }

    std::unique_ptr<TestLelantusDb> CreateDb()
    {
        return std::unique_ptr<TestLelantusDb>(new TestLelantusDb());
    }
};

} // empty namespace

BOOST_FIXTURE_TEST_SUITE(elysium_lelantusdb_tests, LelantusDbTestingSetup)

BOOST_AUTO_TEST_CASE(scalars)
{
    auto db = CreateDb();

#define HAS(id, s) BOOST_CHECK_MESSAGE(db->HasSerial(id, s), \
    strprintf("Expect to have %s in group %d, but not found", secp_primitives::Scalar(s).GetHex(), id))

#define HAS_NO(id, s) BOOST_CHECK_MESSAGE(!db->HasSerial(id, s), \
    strprintf("Expect to have no %s in group %d, but found", secp_primitives::Scalar(s).GetHex(), id))

    HAS_NO(2, 1);
    HAS_NO(3, 2);
    HAS_NO(3, 3);

    BOOST_CHECK_EQUAL(0, db->ReadNextSerialSequence());

    db->WriteSerials(10, {
        {2, {1}},
        {3, {2, 3}}
    });

    BOOST_CHECK_EQUAL(3, db->ReadNextSerialSequence());

    HAS(2, 1);
    HAS(3, 2);

    HAS_NO(2, 2);

    HAS_NO(2, 3);
    HAS_NO(2, 4);

    db->WriteSerials(11, {
        {2, {3, 4}},
    });

    BOOST_CHECK_EQUAL(5, db->ReadNextSerialSequence());

    HAS(2, 3);
    HAS(2, 4);

    db->RemoveSerials(11);

    BOOST_CHECK_EQUAL(3, db->ReadNextSerialSequence());

    HAS_NO(2, 3);
    HAS_NO(2, 4);

    db->RemoveSerials(0);

    HAS_NO(2, 1);
    HAS_NO(3, 2);
    HAS_NO(3, 3);

    BOOST_CHECK_EQUAL(0, db->ReadNextSerialSequence());

#undef HAS
#undef HAS_NO
}

BOOST_AUTO_TEST_SUITE_END()

} // namespace elysium 
 ProTip! Use n and p to navigate between commits in a pull request.
© 2020 GitHub, Inc.
Terms
Privacy
Security
Status
Help
Contact GitHub
Pricing
API
Training
Blog
About

File: qa/rpc-tests/elysium_sendlelantusspend.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        lelantus_starting_block = 1000
+
+        self.nodes[0].generatetoaddress(lelantus_starting_block - self.nodes[0].getblockcount(), self.addrs[0])
+        self.sync_all()
+
+        assert_equal(lelantus_starting_block, self.nodes[0].getblockcount())
+
+        # non-lelantus
+        self.nodes[0].elysium_sendissuancefixed(self.addrs[0], 1, 1, 0, 'main', \
+            'indivisible', 'non-sigma', '', '', '1000000')
```
Comment (nopslide) [inline qa/rpc-tests/elysium_sendlelantusspend.py:26]: Are we using Lelantus or Sigma ?

File: src/Makefile.elysium.include [hunk 2]
```diff
@@ -30,6 +30,8 @@ ELYSIUM_H = \
   elysium/sigma.h \
```
Comment (nopslide) [inline src/Makefile.elysium.include:30]: If we don't use Sigma anymore, I think we should remove it.

File: src/elysium/elysium.cpp [hunk 3]
```diff
@@ -800,6 +809,8 @@ static int parseTransaction(bool bRPConly, const CTransaction& wtx, int nBlock,
     case InputMode::SIGMA:
```
Comment (nopslide) [inline src/elysium/elysium.cpp:809]: I think remove Sigma, and allow only Lelantus will be better.

File: src/elysium/createpayload.cpp [hunk 4]
```diff
@@ -672,5 +682,95 @@ std::vector<unsigned char> CreatePayload_SimpleSpend(
...
+
+    // Mint data
+    {
+        CDataStream serialized(SER_NETWORK, CLIENT_VERSION);
+        serialized << joinSplit;
+        payload.insert(payload.end(), serialized.begin(), serialized.end());
+    }
+
+    // change
+    if (mint.has_value()) {
```
Comment (nopslide) [inline src/elysium/createpayload.cpp:749]: I cannot compile on this line 
`elysium/createpayload.cpp:749:14: error: ‘const class boost::optional<elysium::JoinSplitMint>’ has no member named ‘has_value’; did you mean ‘value’?
     if (mint.has_value()) {
              ^~~~~~~~~
              value`

Comment (nopslide) [reply src/elysium/createpayload.cpp:749]: @thebevrishot

File: src/elysium/rpctx.cpp [hunk 5]
```diff
@@ -1821,6 +1843,238 @@ UniValue elysium_sendspend(const JSONRPCRequest& request)
...
+            reservation->Commit();
+        }
+
+        if (!autoCommit) {
+            return rawHex;
+        } else {
+            PendingAdd(
+                txid,
+                "Spend",
+                ELYSIUM_TYPE_SIMPLE_SPEND,
```
Comment (nopslide) [inline src/elysium/rpctx.cpp:2008]: @thebevrishot Simple Spend is for Sigma ?

File: src/elysium/wallet.cpp [hunk 6]
```diff
@@ -75,12 +92,22 @@ void Wallet::SetSigmaMintUsedTransaction(const SigmaMintId& id, const uint256& t
...
 
+void Wallet::SetLelantusMintUsedTransaction(const MintEntryId& id, const uint256& tx)
+{
+    lelantusWallet.UpdateMintSpendTx(id, tx);
+}
+
 void Wallet::ClearAllChainState()
 {
     mintWalletV0.ClearMintsChainState();
     mintWalletV1.ClearMintsChainState();
```
Comment (nopslide) [inline src/elysium/wallet.cpp:103]: Missing lelantusWallet.ClearMintsChainState(); ?

**Pull request closed without merging**